src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    public MyLinkedList() {
+        this.first = null;
+        this.last = null;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    public void linkFirst(T t) {
+        Node<T> f = first;
+        Node<T> newNode = new Node<>(null, t, first);
+        first = newNode;
+        if (f == null) {
+            last = newNode;
+        } else {
+            f.prev = newNode;
+        }
+        size++;
+    }
+
+    public void linkLast(T t) {
+        if (last == null) {
+            Node<T> newNode = new Node<>(null, t, null);
+            first = newNode;
+            last = newNode;
+        } else {
+            Node<T> newNode = new Node<>(last, t, null);
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Wrong index " + index);
+        }
+        if (index == size) {
+            linkLast(value);
+        } else if (index == 0) {
+            linkFirst(value);
+        } else {
+            Node<T> pred = getNode(index - 1);
+            Node<T> succ = pred.next;
+            Node<T> newNode = new Node<>(pred, value, succ);
+            pred.next = newNode;
+            succ.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Wrong index " + index);
+        }
```

Comment:
include size in the message too
repeated logic, extract to a private utility class